src/webui/api/searchcontroller.cpp

```diff
@@ -0,0 +1,394 @@
+
+    if (!Utils::ForeignApps::pythonInfo().isValid())
+        throw APIError(APIErrorType::Conflict, "Python must be installed to use the Search Engine.");
+
+    const QString pattern = params()["pattern"].trimmed();
+    const QString category = params()["category"].trimmed();
+    const QStringList plugins = params()["plugins"].split('|');
+
+    QStringList pluginsToUse;
+    if (plugins.size() == 1) {
+        const QString pluginsLower = plugins[0].toLower();
+        if (pluginsLower == "all")
+            pluginsToUse = SearchPluginManager::instance()->allPlugins();
+        else if ((pluginsLower == "enabled") || (pluginsLower == "multi"))
+            pluginsToUse = SearchPluginManager::instance()->enabledPlugins();
+        else
+            pluginsToUse << plugins;
+    }
+    else {
+        pluginsToUse << plugins;
+    }
+
+    ISession *session = sessionManager()->session();
+    auto activeSearches = session->getData<QSet<int>>(ACTIVE_SEARCHES);
+    if (activeSearches.size() >= MAX_CONCURRENT_SEARCHES)
+        throw APIError(APIErrorType::Conflict, QString("Unable to create more than %1 concurrent searches.").arg(MAX_CONCURRENT_SEARCHES));
+
+    const auto id = generateSearchId();
+    SearchHandlerPtr searchHandler {SearchPluginManager::instance()->startSearch(pattern, category, pluginsToUse)};
+    QObject::connect(searchHandler.data(), &SearchHandler::searchFinished, this, [session, id, this]() { searchFinished(session, id); });
+    QObject::connect(searchHandler.data(), &SearchHandler::searchFailed, this, [session, id, this]() { searchFailed(session, id); });
+
+    auto searchHandlers = session->getData<SearchHandlerDict>(SEARCH_HANDLERS);
+    searchHandlers.insert(id, searchHandler);
+    session->setData(SEARCH_HANDLERS, QVariant::fromValue(searchHandlers));
+
+    activeSearches.insert(id);
+    session->setData(ACTIVE_SEARCHES, QVariant::fromValue(activeSearches));
+
+    const int hour = 3600000;
+    QTimer::singleShot(hour, this, [id, this] () {searchTimeout(id);});
+
+    const QJsonObject result = {{"id", id}};
+    setResult(result);
+}
+
+void SearchController::stopAction()
+{
+    checkParams({"id"});
+
+    const int id = params()["id"].toInt();
+    ISession *session = sessionManager()->session();
+
+    auto searchHandlers = session->getData<SearchHandlerDict>(SEARCH_HANDLERS);
+    if (!searchHandlers.contains(id))
+        throw APIError(APIErrorType::NotFound);
+
+    SearchHandlerPtr searchHandler = searchHandlers[id];
+
+    if (searchHandler->isActive()) {
+        searchHandler->cancelSearch();
+        removeActiveSearch(session, id);
+    }
+}
+
+void SearchController::statusAction()
+{
+    const int id = params()["id"].toInt();
+
+    auto searchHandlers = sessionManager()->session()->getData<SearchHandlerDict>(SEARCH_HANDLERS);
+    if ((id != 0) && !searchHandlers.contains(id))
+        throw APIError(APIErrorType::NotFound);
+
+    QJsonArray statusArray;
+    QList<int> searchIds;
```

Comment:
I skimmed through, just one last thing from me:
```c++
const QList<int> searchIds {(id == 0) ? searchHandlers.keys() : QList<int> {id}};
// I was wrong, you can omit copyAsConst here
```

Reply:
Can I also omit the `copyAsConst` from line 293: `for (const QString &pluginName : copyAsConst(updateInfo.keys())) {`? `updateInfo` is already const.

Reply:
Yes, but not because it is already const.